### ISSUE TITLE
DbException: ScErrCLRDecToX6DecRangeError (SCERR4246)

### ISSUE THREAD

**Issue by atifwaqar** [Starcounter/Starcounter#4053]
Query which produces this error : 
```
select (sum(product.CampaignPriceSEK) - sum(product.campaignbudgetSEK)) from adfenix.database.campaign
```

Types of attributes : 
public decimal CampaignPriceSEK;
public decimal CampaignBudgetSEK;

```
Starcounter.DbException: ScErrCLRDecToX6DecRangeError (SCERR4246): The CLR Decimal cannot be converted to a Starcounter X6 decimal without data loss. Range error.
Version: 2.2.2.3511.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4246.
at Starcounter.Internal.X6Decimal.ToRaw(Decimal value) in C:\TeamCity\BuildAgent\work\sc-develop-daily-2309\Level1\src\Starcounter.Internal\X6Decimal.cs:line 103
at Starcounter.Query.Execution.ByteArrayBuilder.PrecomputeBuffer(Nullable`1 value, Byte embedType) in C:\TeamCity\BuildAgent\work\sc-develop-daily-2309\Level1\src\Starcounter\Query\Execution\Ranges\ByteArrayBuilder.cs:line 361
at Starcounter.Query.Execution.DecimalSetFunction.GetResult() in C:\TeamCity\BuildAgent\work\sc-develop-daily-2309\Level1\src\Starcounter\Query\Execution\SetFunctions\DecimalSetFunction.cs:line 99
at Starcounter.Query.Execution.Aggregation.CreateNewObject(Row compObject, Boolean produceOneResult) in C:\TeamCity\BuildAgent\work\sc-develop-daily-2309\Level1\src\Starcounter\Query\Execution\Enumerators\Aggregation.cs:line 252
at Starcounter.Query.Execution.Aggregation.MoveNext() in C:\TeamCity\BuildAgent\work\sc-develop-daily-2309\Level1\src\Starcounter\Query\Execution\Enumerators\Aggregation.cs:line 236
at Starcounter.Internal.SqlRestHandler.ExecuteQuery(String query, Int32 maxResult) in C:\TeamCity\BuildAgent\work\sc-develop-daily-2309\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\SqlRestHandler.cs:line 142
ErrorCode=4246
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4246
```

**Comment by mtornwall** [comment]
Probable duplicate of #3841

**Comment by mtornwall** [comment]
If you rewrite the query as `... SUM(product.CampaignPriceSEK - product.CampaignBudgetSEK) ...` (i.e., move the subtraction into the summation) you might be able to avoid overflow.

**Comment by Adeelkh** [comment]
subtraction inside the summation also results in overflow

**Comment by dan31** [comment]
CC @k-rus @chrhol

**Comment by chrhol** [comment]
We also have #4023 which is similar to this and #3841. Maybe it's time to really prioritize or at least take a formal decision if we're going to support X6Decimal or get rid of it since we have quite a few issues around it now where normal usage simply doesn't work.

**Comment by k-rus** [comment]
I checked the source code and it seems that the summation is stored in C# decimal. Thus I don't really understand what exactly happens in the above query and its execution. I will try to reproduce it.

**Comment by dan31** [comment]
@atifwaqar, an issue with ```decimal``` type inconsistency will be addressed in one of later 2.3.x releases as a non-breaking change (but not yet in 2.3 itself). Guys, what workaround we can suggest there while the fix is not present?

**Comment by k-rus** [comment]
> Guys, what workaround we can suggest there while the fix is not present?

I don't really understand why the overflow exception happens here. It needs to be investigated to see what work around can be done.

**Comment by atifwaqar** [comment]
Its ok guys, Do not spend time in fixing this in 2.2.x.
We will be on 2.3.x soon, Till then we can afford this bug.

**Comment by k-rus** [comment]
This bug was fixed in 2.4 during code refactoring. See 52ce5109c13eae630aa9eb9721555f93e852c80a

**Comment by k-rus** [comment]
> This bug was fixed in 2.4 during code refactoring. See 52ce510

Actually this is available in 2.3. I close the issue. If you meet this issue in 2.3. (which is unexpected), then reopen the issue.
